plotly/tools.py

```diff
@@ -1296,3 +1304,686 @@ def return_figure_from_figure_or_data(figure_or_data, validate_figure):
+        v_s = ss/(2*np.pi) * (Y-y_s)/((X-x_s)**2 + (Y-y_s)**2)
+
+        # Create streamline
+        streamline = TraceFactory.create_streamline(x, y, u_s, v_s, density=2,
+                                                    name='streamline')
+
+        # Add source point
+        point = Scatter(x=[x_s], y=[y_s], mode='markers',
+                        marker=Marker(size=14), name='source point')
+        # Plot
+        fig=Figure()
+        fig['data'].append(streamline)
+        fig['data'].append(point)
+        py.iplot(fig, filename='streamline')
+        ```
+        """
+        TraceFactory.validate_streamline(x, y, u, v,
+                                         density, angle,
+                                         arrow_scale, **kwargs)
+        streamline_x, streamline_y = _Streamline(x, y, u, v,
+                                                 density, angle,
+                                                 arrow_scale).sum_streamlines()
+        arrow_x, arrow_y = _Streamline(x, y, u, v,
+                                       density, angle,
+                                       arrow_scale).get_streamline_arrows()
+
+        streamline = Scatter(x=streamline_x + arrow_x,
+                             y=streamline_y + arrow_y,
+                             mode='lines', **kwargs)
+        return streamline
+
+    # def validate():
+
+
+class _Quiver(TraceFactory):
+    def __init__(self, x, y, u, v,
+                 scale, arrow_scale, angle, **kwargs):
+        try:
+            x = self.flatten(x)
+        except exceptions.PlotlyError:
+            pass
+
+        try:
+            y = self.flatten(y)
+        except exceptions.PlotlyError:
+            pass
+
+        try:
+            u = self.flatten(u)
+        except exceptions.PlotlyError:
+            pass
+
+        try:
+            v = self.flatten(v)
+        except exceptions.PlotlyError:
+            pass
+
+        self.x = x
+        self.y = y
+        self.u = u
+        self.v = v
+        self.scale = scale
+        self.arrow_scale = arrow_scale
+        self.angle = angle
+        self.end_x = []
+        self.end_y = []
+        self.validate()
+        self.scale_uv()
+        barb_x, barb_y = self.get_barbs()
+        arrow_x, arrow_y = self.get_quiver_arrows()
+
+    def flatten(self, array):
+        """
+        Uses list comprehension to flatten array
+
+        :param (array): An iterable to flatten
+        :raises (PlotlyError): If iterable is not nested.
+        :rtype (list): The flattened list.
+        """
+        try:
+            return [item for sublist in array for item in sublist]
+        except TypeError:
+            raise exceptions.PlotlyError("Your data array could not be "
+                                         "flattened! Make sure x, y, u, and v "
+                                         "are lists or numpy ndarrays!")
+
+    # todo pull validate method out
```

Comment:
:cow2: todo's should look like this:

`# TODO: pull validate method out`

https://github.com/plotly/dev-docs/blob/466aa0343dd0133eca66ef2e6c562370c106e059/reference/styleguides/python.md#fixme-and-todo